src/main/java/com/rabbitmq/jms/util/WhiteListObjectInputStream.java

```diff
@@ -0,0 +1,242 @@
+        Class[] ifaces = new Class[interfaces.length];
+        for (int i = 0; i < interfaces.length; i++) {
+            ifaces[i] = load(interfaces[i], cl);
+        }
+
+        Class clazz = null;
+        try {
+            clazz = Proxy.getProxyClass(cl, ifaces);
+        } catch (IllegalArgumentException e) {
+            try {
+                clazz = Proxy.getProxyClass(inputStreamLoader, ifaces);
+            } catch (IllegalArgumentException _ignored1) {
+                // ignore
+            }
+            try {
+                clazz = Proxy.getProxyClass(FALLBACK_CLASS_LOADER, ifaces);
+            } catch (IllegalArgumentException _ignored2) {
+                // ignore
+            }
+        }
+
+        if (clazz != null) {
+            checkWhiteList(clazz);
+            return clazz;
+        } else {
+            throw new ClassNotFoundException(null);
+        }
+    }
+
+    @SuppressWarnings("unused")
+    public void addTrustedPackage(String trustedPackage) {
+        this.trustedPackages.add(trustedPackage);
+    }
+
+    /**
+     * @return list of packages trusted for deserialization
+     * from ObjectMessage payloads
+     */
+    @SuppressWarnings("unused")
+    public List<String> getTrustedPackages() {
+        return trustedPackages;
+    }
+
+    /**
+     * @param trustedPackages list of packages trusted for deserialization
+     *                        from ObjectMessage payloads
+     */
+    @SuppressWarnings("unused")
+    public void setTrustedPackages(List<String> trustedPackages) {
+        this.trustedPackages = trustedPackages;
+    }
+
+    /**
+     * @return true if this object stream considers all packages to
+     * be trusted, false otherwise
+     */
+    public boolean shouldTrustAllPackages() {
+        return (this.trustedPackages != null) && (trustedPackages.size() == 1 && trustedPackages.get(0).equals("*"));
+    }
+
+    private void checkWhiteList(Class clazz) throws ClassNotFoundException {
+        if (clazz.isPrimitive()) {
+            return;
+        }
+
+        if (clazz.getPackage() != null && !shouldTrustAllPackages()) {
+            boolean result = false;
+            String p = clazz.getPackage().getName();
+            for (String pkg : this.trustedPackages) {
+                // Note: this means that an empty string works the same way as "*"
+                //       but making it mean "trust no package" makes even less sense
+                if (p.equals(pkg) || p.startsWith(pkg)) {
+                    result = true;
+                    break;
+                }
+            }
+
+            if (!result) {
+                throw new ClassNotFoundException(
+                                                        "Class " + clazz + " is not trusted to be serialized as ObjectMessage payload. "
+                                                                + "Trusted packages can be configured via RMQConnectionFactory.");
```

Comment:
@michaelklishin , I'm not sure why we are talking about `serialization` here, when it seems for me that the real problem is exactly on the `de-serialization`, when we receive `ObjectMessage`. But I see similar message in the ActiveMQ, too:
```
throw new ClassNotFoundException("Forbidden " + clazz + "! This class is not trusted to be serialized as ObjectMessage payload. Please take a look at http://activemq.apache.org/objectmessage.html for more information on how to configure trusted classes.");
```
Meanwhile @garyrussell says like:
```
throw new SecurityException("Attempt to deserialize unauthorized " + clazz);
```

If I understand correctly this is the main place for review. So, looks good, of course.

It's so pity that you have mixed with this fix other polishing, which make the review a bit complicated and I can't be sure in all other changes.

Reviewing further...